### ISSUE TITLE
ARC: "samples/logger-hook" crashed __memory_error from sys_ring_buf_get

### ISSUE THREAD

**Issue by nashif** [zephyriot/zep-jira14#1639]
**_Reported by Sharron LIU:_**

When running "zephyr/samples/logger-hook" at arduino_101_sss (ARC), software crashed at "sys_ring_buf_get".
Here's the info shown by GDB:
```
(gdb) b __memory_error
Breakpoint 1 at 0x40035ba4: file /home/sharron/workspace/views/iot/zephyr/arch/arc/core/fault_s.S, line 90.
(gdb) c
Continuing.

Breakpoint 1, __memory_error () at /home/sharron/workspace/views/iot/zephyr/arch/arc/core/fault_s.S:90
90              mov_s sp, _firq_stack
(gdb) bt
<span>#</span>0  __memory_error () at /home/sharron/workspace/views/iot/zephyr/arch/arc/core/fault_s.S:90
<span>#</span>1  0x400343d0 in ring_buf_print (buf=<optimized out>) at /home/sharron/workspace/views/iot/zephyr/samples/logger-hook/src/main.c:86
<span>#</span>2  0x40036aa4 in __config_sys_init_quark_se_arc_init0 ()
Backtrace stopped: frame did not save the PC
(gdb) f 1
<span>#</span>1  0x400343d0 in ring_buf_print (buf=<optimized out>) at /home/sharron/workspace/views/iot/zephyr/samples/logger-hook/src/main.c:86
86                      ret = sys_ring_buf_get(&log_cbuffer.ring_buffer, 0, 0,
(gdb) p log_cbuffer.ring_buffer
$1 = {head = 0, tail = 52, dropped_put_count = 0, size = 512, buf = 0xa80005a0 <logger_buffer>, mask = 511}
```

how-to-reproduce:
# source zephyr-env.sh
# cd samples/logger-hook
# make BOARD=arduino_101_sss flash

(Imported from Jira ZEP-1789)

**Comment by nashif** [comment]
**_by Andrew Boie:_**

The ring buffer data structure was being mis-used in several ways. Patch is up to fix it.

**Comment by nashif** [comment]
**_by Mark Linkmeyer:_**

Hi _Sharron LIU_ , I assume you've seen that this bug needs verification by QA.  Thx.

**Comment by nashif** [comment]
**_by Sharron LIU:_**

Verified fixed at master while patch not in v1.7 RC4.
